### PR TITLE
Queue counter

### PR DESCRIPTION
**Pull request by ohaz**

### CODE REVIEW

File: backend/app.py [hunk 1]
```diff
@@ -1,12 +1,20 @@
...
-def hello_world():
-    return 'Hey, we have Flask in a Docker container!'
+app = Flask(__name__)
 
+app.config['SQLALCHEMY_DATABASE_URI'] = f'sqlite:///{SQLITE_FILE}'
+db = SQLAlchemy(app)
 
-if __name__ == '__main__':
-    app.run(debug=True, host='0.0.0.0')
+if not isfile(SQLITE_FILE):
```
Comment (fnbr) [inline backend/app.py]: use Path.is_file() so we don't mix two path libraries

File: backend/app.py [hunk 2]
```diff
@@ -1,12 +1,20 @@
...
+    Path(SQLITE_FILE).parent.absolute().mkdir(parents=True, exist_ok=True)
+    from models.slot import Slot
+    from models.queue import Queue
+    db.create_all()
+    db.session.commit()
```
Comment (fnbr) [inline backend/app.py]: no newline at end of file

File: backend/digitales_warten.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    return 'Hey, we have Flask in a Docker container!'
+
+@app.route('/queue', methods=['POST'])
+def create_queue():
+    queue = Queue()
+    db.session.add(queue)
+    db.session.commit()
+    return str(queue.id)
+
+@app.route('/slot/<queue_id>', methods=['POST'])
```
Comment (fnbr) [inline backend/digitales_warten.py]: I know nothing about web-stuff, but this path look strange. having /slot/42 looks like 42 is the id of the slot. I would have expected something like /queue/42/slot

File: backend/run_tests.sh [hunk 4]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+docker-compose up --abort-on-container-exit --build --force-recreate
```
Comment (Gronner) [inline backend/run_tests.sh]: Missing empty line at file end!

File: backend/digitales_warten.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+def print_slot(queue_id):
```
Comment (fnbr) [inline backend/digitales_warten.py]: why print? looks like you create a slot

File: backend/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+            context: .
+            dockerfile: Dockerfile
+        ports:
+            - 5000:5000
+    tests:
+        build:
+            context: .
+            dockerfile: tests.Dockerfile
+        depends_on:
+            - backend
```
Comment (fnbr) [inline backend/docker-compose.yml]: nl@eof

File: backend/docker.sh [hunk 7]
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+docker build -t dwb .
+#docker run --volume /$(pwd)/db:/db -p 5000:5000 dwb
+docker run -p 5000:5000 dwb
```
Comment (fnbr) [inline backend/docker.sh]: nl@eof

File: backend/models/queue.py [hunk 8]
```diff
@@ -0,0 +1,7 @@
+from app import db
+
+class Queue(db.Model):
+    id = db.Column(db.Integer, primary_key=True)
+
+    def __repr__(self):
+        return f'<Queue {self.id}: [{self.slots}]>'
```
Comment (fnbr) [inline backend/models/queue.py]: nl@eof

File: backend/models/slot.py [hunk 9]
```diff
@@ -0,0 +1,9 @@
+from app import db
+
+class Slot(db.Model):
+    id = db.Column(db.Integer, primary_key=True)
+    queue_id = db.Column(db.Integer, db.ForeignKey('queue.id'))
+    queue = db.relationship('Queue', backref=db.backref('slots'))
+
+    def __repr__(self):
+        return f'<Slot {id}>'
```
Comment (fnbr) [inline backend/models/slot.py]: nl@eof

File: backend/run_tests.sh (hunk 4 above)
Comment (fnbr) [inline backend/run_tests.sh]: nl@eof

Comment (ohaz) [reply backend/run_tests.sh]: Done

File: backend/tests/requirements.txt [hunk 10]
```diff
@@ -0,0 +1,2 @@
+pytest==5.4.1
+requests==2.23.0
```
Comment (fnbr) [inline backend/tests/requirements.txt]: A setup.cfg with different variants/features might be better than having multiple requirements.txt. But we can refactor that later

File: backend/app.py (hunk 1 above)
Comment (ohaz) [reply backend/app.py]: Done

File: backend/app.py (hunk 2 above)
Comment (ohaz) [reply backend/app.py]: Done

File: backend/docker-compose.yml (hunk 6 above)
Comment (ohaz) [reply backend/docker-compose.yml]: Done

File: backend/docker.sh (hunk 7 above)
Comment (ohaz) [reply backend/docker.sh]: Done

File: backend/models/slot.py (hunk 9 above)
Comment (ohaz) [reply backend/models/slot.py]: Done

File: backend/run_tests.sh (hunk 4 above)
Comment (ohaz) [reply backend/run_tests.sh]: Done

File: backend/digitales_warten.py (hunk 5 above)
Comment (ohaz) [reply backend/digitales_warten.py]: Done

File: backend/digitales_warten.py (hunk 3 above)
Comment (ohaz) [reply backend/digitales_warten.py]: I know nothing about web-stuff either, so I just used the first thing I thought of. We should have an API meeting with frontend about that, they probably know more about it :)

File: backend/app.py [hunk 11]
```diff
@@ -1,12 +1,19 @@
...
-    return 'Hey, we have Flask in a Docker container!'
+app = Flask(__name__)
 
+app.config['SQLALCHEMY_DATABASE_URI'] = f'sqlite:///{SQLITE_FILE}'
+db = SQLAlchemy(app)
 
-if __name__ == '__main__':
-    app.run(debug=True, host='0.0.0.0')
+if not Path.is_file(SQLITE_FILE):
+    Path(SQLITE_FILE).parent.absolute().mkdir(parents=True, exist_ok=True)
```
Comment (fnbr) [inline backend/app.py]:
```suggestion
db_file = Path(SQLITE_FILE)
if not db_file.is_file():
    db_file.parent.absolute().mkdir(parents=True, exist_ok=True)
```
Or just leave out the

```suggestion
Path(SQLITE_FILE).parent.absolute().mkdir(parents=True, exist_ok=True)
```
mkdir should be file anyways

Comment (ohaz) [reply backend/app.py]: I just tested it, when running `Path('a/b.foo').mkdir(parents=True)` a folder named `b.foo` is created

File: backend/digitales_warten.py (hunk 3 above)
Comment (ohaz) [reply backend/digitales_warten.py]: Done

**Pull request merged**